File: codility/src/main/java/codility/com/lessons/l09maxsliceproblem/MaxSliceSum.java
package codility.com.lessons.l09maxsliceproblem;

/**
 * 100%
 * https://app.codility.com/programmers/lessons/9-maximum_slice_problem/max_slice_sum/
 */
public class MaxSliceSum {
	public int solution(int[] A) {
		if (A.length == 1) {
			return A[0];
		}

		int max_ending = 0;
		int max_slice = 0;

		for (int a : A) {
			max_ending = Math.max(0, max_ending + a);
			max_slice = Math.max(max_slice, max_ending);
		}

		if (max_slice == 0) {
			max_slice = A[0];
			for (int a : A) {
				max_slice = Math.max(max_slice, a);
			}
		}
		return max_slice;
	}
}

File: codility_new/max_slice_sum.py
# https://app.codility.com/programmers/lessons/9-maximum_slice_problem/max_slice_sum/

def solution(A):
    if len(A) == 0:
        return 0
    if len(A) == 1:
        return A[0]
    max_ending = A[0]
    max_slice = A[0]
    for i in range(1, len(A)):
        a = A[i]
        max_ending = max(a, max_ending + a)
        max_slice = max(max_slice, max_ending)
    return max_slice
